
<!doctype html>
<html lang="en" class="no-js">
  <head>

      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">

        <meta name="description" content="G00x: A generalized pipeline for analyzing germline-targeting clinical trials">



        <link rel="canonical" href="https://g00x.schieflab.com/g003_pipeline/">


        <link rel="prev" href="../g003_validation/">


        <link rel="next" href="../g003_analysis/">


      <link rel="icon" href="../img/favico.svg">
      <meta name="generator" content="mkdocs-1.6.1, mkdocs-material-9.6.12">



        <title>Pipeline - G00x</title>



      <link rel="stylesheet" href="../assets/stylesheets/main.2afb09e1.min.css">


        <link rel="stylesheet" href="../assets/stylesheets/palette.06af60db.min.css">








  <style>:root{--md-admonition-icon--info:url('data:image/svg+xml;charset=utf-8,%3Csvg%20xmlns%3D%22http%3A//www.w3.org/2000/svg%22%20viewBox%3D%220%200%2024%2024%22%3E%3Cpath%20d%3D%22M7%202v2h1v14a4%204%200%200%200%204%204%204%204%200%200%200%204-4V4h1V2zm4%2014c-.6%200-1-.4-1-1s.4-1%201-1%201%20.4%201%201-.4%201-1%201m2-4c-.6%200-1-.4-1-1s.4-1%201-1%201%20.4%201%201-.4%201-1%201m1-5h-4V4h4z%22/%3E%3C/svg%3E');}</style>











        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>



      <link rel="stylesheet" href="../css/termynal.css">

      <link rel="stylesheet" href="../css/custom.css">

    <script>__md_scope=new URL("..",location),__md_hash=e=>[...e].reduce(((e,_)=>(e<<5)-e+_.charCodeAt(0)),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>









<meta property="og:type" content="website" />


<meta property="og:url" content="https://g00x.schieflab.com/g003_pipeline/" />


<meta property="og:image:url" content="" />



<meta property="og:image:type" content="image/png" />



<meta property="og:description" content="G00x: A generalized pipeline for analyzing germline-targeting clinical trials">



  </head>









    <body dir="ltr" data-md-color-scheme="default" data-md-color-primary="white" data-md-color-accent="amber">


    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">


        <a href="#10x-pipeline" class="md-skip">
          Skip to content
        </a>

    </div>
    <div data-md-component="announce">

    </div>






<header class="md-header md-header--shadow md-header--lifted" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href=".." title="G00x" class="md-header__button md-logo" aria-label="G00x" data-md-component="logo">


  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="m11.15 15.18-1.42-1.41 1.42-1.42 1.41 1.42 1.41-1.42-1.41-1.41 1.41-1.41 1.42 1.41 1.41-1.41-2.83-2.83-7.07 7.07 2.83 2.83zM3.08 19l3.12-3.11-2.12-2.12 9.89-9.9L16.1 6l1.4-1.42-1.4-1.42 1.4-1.41L21.75 6l-1.41 1.4L18.92 6 17.5 7.4l2.13 2.13-9.9 9.89-2.12-2.12-4.53 4.54z"/></svg>

    </a>
    <label class="md-header__button md-icon" for="__drawer">

      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3zm0 5h18v2H3zm0 5h18v2H3z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            G00x
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">

              Pipeline

          </span>
        </div>
      </div>
    </div>


        <form class="md-header__option" data-md-component="palette">




    <input class="md-option" data-md-color-media="" data-md-color-scheme="default" data-md-color-primary="white" data-md-color-accent="amber"  aria-label="Switch to night mode"  type="radio" name="__palette" id="__palette_0">

      <label class="md-header__button md-icon" title="Switch to night mode" for="__palette_1" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="m17.75 4.09-2.53 1.94.91 3.06-2.63-1.81-2.63 1.81.91-3.06-2.53-1.94L12.44 4l1.06-3 1.06 3zm3.5 6.91-1.64 1.25.59 1.98-1.7-1.17-1.7 1.17.59-1.98L15.75 11l2.06-.05L18.5 9l.69 1.95zm-2.28 4.95c.83-.08 1.72 1.1 1.19 1.85-.32.45-.66.87-1.08 1.27C15.17 23 8.84 23 4.94 19.07c-3.91-3.9-3.91-10.24 0-14.14.4-.4.82-.76 1.27-1.08.75-.53 1.93.36 1.85 1.19-.27 2.86.69 5.83 2.89 8.02a9.96 9.96 0 0 0 8.02 2.89m-1.64 2.02a12.08 12.08 0 0 1-7.8-3.47c-2.17-2.19-3.33-5-3.49-7.82-2.81 3.14-2.7 7.96.31 10.98 3.02 3.01 7.84 3.12 10.98.31"/></svg>
      </label>





    <input class="md-option" data-md-color-media="" data-md-color-scheme="slate" data-md-color-primary="black" data-md-color-accent="red"  aria-label="Switch to light mode"  type="radio" name="__palette" id="__palette_1">

      <label class="md-header__button md-icon" title="Switch to light mode" for="__palette_0" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 7a5 5 0 0 1 5 5 5 5 0 0 1-5 5 5 5 0 0 1-5-5 5 5 0 0 1 5-5m0 2a3 3 0 0 0-3 3 3 3 0 0 0 3 3 3 3 0 0 0 3-3 3 3 0 0 0-3-3m0-7 2.39 3.42C13.65 5.15 12.84 5 12 5s-1.65.15-2.39.42zM3.34 7l4.16-.35A7.2 7.2 0 0 0 5.94 8.5c-.44.74-.69 1.5-.83 2.29zm.02 10 1.76-3.77a7.131 7.131 0 0 0 2.38 4.14zM20.65 7l-1.77 3.79a7.02 7.02 0 0 0-2.38-4.15zm-.01 10-4.14.36c.59-.51 1.12-1.14 1.54-1.86.42-.73.69-1.5.83-2.29zM12 22l-2.41-3.44c.74.27 1.55.44 2.41.44.82 0 1.63-.17 2.37-.44z"/></svg>
      </label>


</form>



      <script>var palette=__md_get("__palette");if(palette&&palette.color){if("(prefers-color-scheme)"===palette.color.media){var media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']");palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent")}for(var[key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script>





        <label class="md-header__button md-icon" for="__search">

          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        </label>
        <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">

        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>

        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">

          <a href="javascript:void(0)" class="md-search__icon md-icon" title="Share" aria-label="Share" data-clipboard data-clipboard-text="" data-md-component="search-share" tabindex="-1">

            <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M18 16.08c-.76 0-1.44.3-1.96.77L8.91 12.7c.05-.23.09-.46.09-.7s-.04-.47-.09-.7l7.05-4.11c.54.5 1.25.81 2.04.81a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3c0 .24.04.47.09.7L8.04 9.81C7.5 9.31 6.79 9 6 9a3 3 0 0 0-3 3 3 3 0 0 0 3 3c.79 0 1.5-.31 2.04-.81l7.12 4.15c-.05.21-.08.43-.08.66 0 1.61 1.31 2.91 2.92 2.91s2.92-1.3 2.92-2.91A2.92 2.92 0 0 0 18 16.08"/></svg>
          </a>

        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">

          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"/></svg>
        </button>
      </nav>

        <div class="md-search__suggest" data-md-component="search-suggest"></div>

    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" tabindex="0" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>



      <div class="md-header__source">
        <a href="https://github.com/schieflab/g00x" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">

    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 480 512"><!--! Font Awesome Free 6.7.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M186.1 328.7c0 20.9-10.9 55.1-36.7 55.1s-36.7-34.2-36.7-55.1 10.9-55.1 36.7-55.1 36.7 34.2 36.7 55.1M480 278.2c0 31.9-3.2 65.7-17.5 95-37.9 76.6-142.1 74.8-216.7 74.8-75.8 0-186.2 2.7-225.6-74.8-14.6-29-20.2-63.1-20.2-95 0-41.9 13.9-81.5 41.5-113.6-5.2-15.8-7.7-32.4-7.7-48.8 0-21.5 4.9-32.3 14.6-51.8 45.3 0 74.3 9 108.8 36 29-6.9 58.8-10 88.7-10 27 0 54.2 2.9 80.4 9.2 34-26.7 63-35.2 107.8-35.2 9.8 19.5 14.6 30.3 14.6 51.8 0 16.4-2.6 32.7-7.7 48.2 27.5 32.4 39 72.3 39 114.2m-64.3 50.5c0-43.9-26.7-82.6-73.5-82.6-18.9 0-37 3.4-56 6-14.9 2.3-29.8 3.2-45.1 3.2-15.2 0-30.1-.9-45.1-3.2-18.7-2.6-37-6-56-6-46.8 0-73.5 38.7-73.5 82.6 0 87.8 80.4 101.3 150.4 101.3h48.2c70.3 0 150.6-13.4 150.6-101.3m-82.6-55.1c-25.8 0-36.7 34.2-36.7 55.1s10.9 55.1 36.7 55.1 36.7-34.2 36.7-55.1-10.9-55.1-36.7-55.1"/></svg>
  </div>
  <div class="md-source__repository">
    schieflab/g00x
  </div>
</a>
      </div>

  </nav>



</header>

    <div class="md-container" data-md-component="container">




      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">



              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">




<nav class="md-nav md-nav--primary" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href=".." title="G00x" class="md-nav__button md-logo" aria-label="G00x" data-md-component="logo">


  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="m11.15 15.18-1.42-1.41 1.42-1.42 1.41 1.42 1.41-1.42-1.41-1.41 1.41-1.41 1.42 1.41 1.41-1.41-2.83-2.83-7.07 7.07 2.83 2.83zM3.08 19l3.12-3.11-2.12-2.12 9.89-9.9L16.1 6l1.4-1.42-1.4-1.42 1.4-1.41L21.75 6l-1.41 1.4L18.92 6 17.5 7.4l2.13 2.13-9.9 9.89-2.12-2.12-4.53 4.54z"/></svg>

    </a>
    G00x
  </label>

    <div class="md-nav__source">
      <a href="https://github.com/schieflab/g00x" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">

    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 480 512"><!--! Font Awesome Free 6.7.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M186.1 328.7c0 20.9-10.9 55.1-36.7 55.1s-36.7-34.2-36.7-55.1 10.9-55.1 36.7-55.1 36.7 34.2 36.7 55.1M480 278.2c0 31.9-3.2 65.7-17.5 95-37.9 76.6-142.1 74.8-216.7 74.8-75.8 0-186.2 2.7-225.6-74.8-14.6-29-20.2-63.1-20.2-95 0-41.9 13.9-81.5 41.5-113.6-5.2-15.8-7.7-32.4-7.7-48.8 0-21.5 4.9-32.3 14.6-51.8 45.3 0 74.3 9 108.8 36 29-6.9 58.8-10 88.7-10 27 0 54.2 2.9 80.4 9.2 34-26.7 63-35.2 107.8-35.2 9.8 19.5 14.6 30.3 14.6 51.8 0 16.4-2.6 32.7-7.7 48.2 27.5 32.4 39 72.3 39 114.2m-64.3 50.5c0-43.9-26.7-82.6-73.5-82.6-18.9 0-37 3.4-56 6-14.9 2.3-29.8 3.2-45.1 3.2-15.2 0-30.1-.9-45.1-3.2-18.7-2.6-37-6-56-6-46.8 0-73.5 38.7-73.5 82.6 0 87.8 80.4 101.3 150.4 101.3h48.2c70.3 0 150.6-13.4 150.6-101.3m-82.6-55.1c-25.8 0-36.7 34.2-36.7 55.1s10.9 55.1 36.7 55.1 36.7-34.2 36.7-55.1-10.9-55.1-36.7-55.1"/></svg>
  </div>
  <div class="md-source__repository">
    schieflab/g00x
  </div>
</a>
    </div>

  <ul class="md-nav__list" data-md-scrollfix>







    <li class="md-nav__item">
      <a href=".." class="md-nav__link">



  <span class="md-ellipsis">
    Introduction

  </span>


      </a>
    </li>














    <li class="md-nav__item md-nav__item--nested">



        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_2" >


          <label class="md-nav__link" for="__nav_2" id="__nav_2_label" tabindex="0">



  <span class="md-ellipsis">
    G002

  </span>


            <span class="md-nav__icon md-icon"></span>
          </label>

        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_2_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_2">
            <span class="md-nav__icon md-icon"></span>
            G002
          </label>
          <ul class="md-nav__list" data-md-scrollfix>







    <li class="md-nav__item">
      <a href="../g002_data/" class="md-nav__link">



  <span class="md-ellipsis">
    Data

  </span>


      </a>
    </li>










    <li class="md-nav__item">
      <a href="../g002_validation/" class="md-nav__link">



  <span class="md-ellipsis">
    Validation

  </span>


      </a>
    </li>










    <li class="md-nav__item">
      <a href="../g002_pipeline/" class="md-nav__link">



  <span class="md-ellipsis">
    Pipeline

  </span>


      </a>
    </li>










    <li class="md-nav__item">
      <a href="../g002_analysis/" class="md-nav__link">



  <span class="md-ellipsis">
    Analysis and Visualization

  </span>


      </a>
    </li>










    <li class="md-nav__item">
      <a href="../g002_cloud/" class="md-nav__link">



  <span class="md-ellipsis">
    Cloud Management (Advanced)

  </span>


      </a>
    </li>




          </ul>
        </nav>

    </li>
















    <li class="md-nav__item md-nav__item--active md-nav__item--nested">



        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_3" checked>


          <label class="md-nav__link" for="__nav_3" id="__nav_3_label" tabindex="0">



  <span class="md-ellipsis">
    G003

  </span>


            <span class="md-nav__icon md-icon"></span>
          </label>

        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_3_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_3">
            <span class="md-nav__icon md-icon"></span>
            G003
          </label>
          <ul class="md-nav__list" data-md-scrollfix>







    <li class="md-nav__item">
      <a href="../g003_data/" class="md-nav__link">



  <span class="md-ellipsis">
    Data

  </span>


      </a>
    </li>










    <li class="md-nav__item">
      <a href="../g003_validation/" class="md-nav__link">



  <span class="md-ellipsis">
    Validation

  </span>


      </a>
    </li>












    <li class="md-nav__item md-nav__item--active">

      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">





        <label class="md-nav__link md-nav__link--active" for="__toc">



  <span class="md-ellipsis">
    Pipeline

  </span>


          <span class="md-nav__icon md-icon"></span>
        </label>

      <a href="./" class="md-nav__link md-nav__link--active">



  <span class="md-ellipsis">
    Pipeline

  </span>


      </a>



<nav class="md-nav md-nav--secondary" aria-label="Table of contents">






    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>

        <li class="md-nav__item">
  <a href="#flow" class="md-nav__link">
    <span class="md-ellipsis">
      Flow
    </span>
  </a>

</li>

        <li class="md-nav__item">
  <a href="#demultiplexing" class="md-nav__link">
    <span class="md-ellipsis">
      Demultiplexing
    </span>
  </a>

</li>

        <li class="md-nav__item">
  <a href="#vdj" class="md-nav__link">
    <span class="md-ellipsis">
      VDJ
    </span>
  </a>

</li>

        <li class="md-nav__item">
  <a href="#cso" class="md-nav__link">
    <span class="md-ellipsis">
      CSO
    </span>
  </a>

</li>

        <li class="md-nav__item">
  <a href="#airr" class="md-nav__link">
    <span class="md-ellipsis">
      AIRR
    </span>
  </a>

</li>

    </ul>

</nav>

    </li>










    <li class="md-nav__item">
      <a href="../g003_analysis/" class="md-nav__link">



  <span class="md-ellipsis">
    Analysis and Visualization

  </span>


      </a>
    </li>




          </ul>
        </nav>

    </li>









    <li class="md-nav__item">
      <a href="../Contributing/" class="md-nav__link">



  <span class="md-ellipsis">
    Contributing

  </span>


      </a>
    </li>



  </ul>
</nav>
                  </div>
                </div>
              </div>



              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">


<nav class="md-nav md-nav--secondary" aria-label="Table of contents">






    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>

        <li class="md-nav__item">
  <a href="#flow" class="md-nav__link">
    <span class="md-ellipsis">
      Flow
    </span>
  </a>

</li>

        <li class="md-nav__item">
  <a href="#demultiplexing" class="md-nav__link">
    <span class="md-ellipsis">
      Demultiplexing
    </span>
  </a>

</li>

        <li class="md-nav__item">
  <a href="#vdj" class="md-nav__link">
    <span class="md-ellipsis">
      VDJ
    </span>
  </a>

</li>

        <li class="md-nav__item">
  <a href="#cso" class="md-nav__link">
    <span class="md-ellipsis">
      CSO
    </span>
  </a>

</li>

        <li class="md-nav__item">
  <a href="#airr" class="md-nav__link">
    <span class="md-ellipsis">
      AIRR
    </span>
  </a>

</li>

    </ul>

</nav>
                  </div>
                </div>
              </div>



            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">








<h1 id="10x-pipeline">10x pipeline<a class="headerlink" href="#10x-pipeline" title="Permanent link">&para;</a></h1>
<p>After you have a validated merged dataframe in from <a href="../g003_validation/">validation</a>, you can begin the 10X pipeline which consists of demultiplexing, vdj, and feature counting.</p>
<p><img alt="full pipeline" src="../img/G003_Pipeline.png" /></p>
<div class="admonition warning">
<p class="admonition-title">CellRanger 7.1.0</p>
<p>The cellranger version in this pipeline is 7.1.0</p>
</div>
<h2 id="flow">Flow<a class="headerlink" href="#flow" title="Permanent link">&para;</a></h2>
<div class="tabbed-set tabbed-alternate" data-tabs="1:2"><input checked="checked" id="__tabbed_1_1" name="__tabbed_1" type="radio" /><input id="__tabbed_1_2" name="__tabbed_1" type="radio" /><div class="tabbed-labels"><label for="__tabbed_1_1">:material-console-line: Command Line Usage</label><label for="__tabbed_1_2"> :material-api: Python</label></div>
<div class="tabbed-content">
<div class="tabbed-block">
<p>The following will analyze the flow path and output a dataframe that will be used in analysis. It computes each count of each gate that are useful in frequency analysis.</p>
<p><div class="termy">
<div class="highlight"><pre><span></span><code>$<span class="w"> </span>g00x<span class="w"> </span>g003<span class="w"> </span>pipeline<span class="w"> </span>flow<span class="w"> </span>-o<span class="w"> </span>g003/G003/output/flow<span class="w"> </span>/path/to/flow
</code></pre></div>
</div></p>
</div>
<div class="tabbed-block">
<div class="highlight"><pre><span></span><code><span class="kn">import</span><span class="w"> </span><span class="nn">pandas</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">pd</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">g00x.data</span><span class="w"> </span><span class="kn">import</span> <span class="n">Data</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">g00x.flow.flow</span><span class="w"> </span><span class="kn">import</span> <span class="n">parse_flow_data</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">pathlib</span><span class="w"> </span><span class="kn">import</span> <span class="n">Path</span>

<span class="n">data</span> <span class="o">=</span> <span class="n">ctx</span><span class="o">.</span><span class="n">obj</span><span class="p">[</span><span class="s2">&quot;data&quot;</span><span class="p">]</span>
<span class="n">folder</span> <span class="o">=</span> <span class="s1">&#39;path/to/flow&#39;</span>
<span class="n">flow_data</span> <span class="o">=</span> <span class="n">parse_flow_data</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">folder</span><span class="p">)</span>
<span class="n">out</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="n">out</span><span class="p">)</span>
<span class="n">output_feather</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="n">out</span><span class="o">.</span><span class="n">parent</span> <span class="o">/</span> <span class="p">(</span><span class="n">out</span><span class="o">.</span><span class="n">stem</span> <span class="o">+</span> <span class="s2">&quot;.feather&quot;</span><span class="p">))</span>
<span class="n">output_csv</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="n">out</span><span class="o">.</span><span class="n">parent</span> <span class="o">/</span> <span class="p">(</span><span class="n">out</span><span class="o">.</span><span class="n">stem</span> <span class="o">+</span> <span class="s2">&quot;.csv&quot;</span><span class="p">))</span>
</code></pre></div>
</div>
</div>
</div>
<p>Here is what the flow dataframe will look like.</p>
<table>
<thead>
<tr>
<th style="text-align: right;"></th>
<th style="text-align: left;">run_purpose</th>
<th style="text-align: left;">run_date</th>
<th style="text-align: left;">sort_id</th>
<th style="text-align: left;">ptid</th>
<th style="text-align: right;">group</th>
<th style="text-align: right;">weeks</th>
<th style="text-align: left;">visit_id</th>
<th style="text-align: left;">probe_set</th>
<th style="text-align: left;">sample_type</th>
<th style="text-align: left;">sort_software_dv</th>
<th style="text-align: left;">sort_file_type</th>
<th style="text-align: left;">sample_tube</th>
<th style="text-align: left;">gate</th>
<th style="text-align: left;">phenotype</th>
<th style="text-align: left;">value_type</th>
<th style="text-align: left;">extention</th>
<th style="text-align: left;">file_path</th>
<th style="text-align: left;">file_subset</th>
<th style="text-align: right;">value</th>
<th style="text-align: left;">branch</th>
<th style="text-align: left;">easy_name</th>
<th style="text-align: right;">notes</th>
<th style="text-align: left;">sort_pool</th>
<th style="text-align: left;">hashtag</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align: right;">119</td>
<td style="text-align: left;">PreS</td>
<td style="text-align: left;">2022-08-25</td>
<td style="text-align: left;">S6C</td>
<td style="text-align: left;">G003831</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">-5</td>
<td style="text-align: left;">V091</td>
<td style="text-align: left;">eODGT8</td>
<td style="text-align: left;">PBMC</td>
<td style="text-align: left;">DV</td>
<td style="text-align: left;">Summary</td>
<td style="text-align: left;">T1</td>
<td style="text-align: left;">P11</td>
<td style="text-align: left;">IgD+/Antigen++</td>
<td style="text-align: left;">count</td>
<td style="text-align: left;">.csv</td>
<td style="text-align: left;">['g003/G003/sorting/G003/Prescreens/Prescreen_RunDate220825_UploadDate221021/PopulationSummaryFilesFromDV/PreS_220825_S6C_G003831_V091_eODGT8_PBMC_DV_Summary_T1_a.csv']</td>
<td style="text-align: left;">['a']</td>
<td style="text-align: right;">50</td>
<td style="text-align: left;">IgD+</td>
<td style="text-align: left;">antigen_pos_igd_pos_b_cells</td>
<td style="text-align: right;">nan</td>
<td style="text-align: left;"></td>
<td style="text-align: left;"></td>
</tr>
<tr>
<td style="text-align: right;">2461</td>
<td style="text-align: left;">PreS</td>
<td style="text-align: left;">2022-11-04</td>
<td style="text-align: left;">S6C</td>
<td style="text-align: left;">G003136</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">8</td>
<td style="text-align: left;">V200</td>
<td style="text-align: left;">eODGT8</td>
<td style="text-align: left;">PBMC</td>
<td style="text-align: left;">DV</td>
<td style="text-align: left;">Summary</td>
<td style="text-align: left;">T1</td>
<td style="text-align: left;">P13</td>
<td style="text-align: left;">IgD+/KO-</td>
<td style="text-align: left;">count</td>
<td style="text-align: left;">.csv</td>
<td style="text-align: left;">['g003/G003/sorting/G003/Prescreens/Prescreen_RunDate221104_UploadDate221129/PopulationSummaryFilesFromDV/PreS_221104_S6C_G003136_V200_eODGT8_PBMC_DV_Summary_T1_a.csv']</td>
<td style="text-align: left;">['a']</td>
<td style="text-align: right;">190682</td>
<td style="text-align: left;">IgD+</td>
<td style="text-align: left;"></td>
<td style="text-align: right;">nan</td>
<td style="text-align: left;"></td>
<td style="text-align: left;"></td>
</tr>
<tr>
<td style="text-align: right;">2754</td>
<td style="text-align: left;">PreS</td>
<td style="text-align: left;">2022-11-04</td>
<td style="text-align: left;">S6C</td>
<td style="text-align: left;">G003947</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">8</td>
<td style="text-align: left;">V200</td>
<td style="text-align: left;">Cg28v2</td>
<td style="text-align: left;">PBMC</td>
<td style="text-align: left;">DV</td>
<td style="text-align: left;">Summary</td>
<td style="text-align: left;">T1</td>
<td style="text-align: left;">P31</td>
<td style="text-align: left;">IgG-IgM-/IgA+/KO-</td>
<td style="text-align: left;">count</td>
<td style="text-align: left;">.csv</td>
<td style="text-align: left;">['g003/G003/sorting/G003/Prescreens/Prescreen_RunDate221104_UploadDate221129/PopulationSummaryFilesFromDV/PreS_221104_S6C_G003947_V200_Cg28v2_PBMC_DV_Summary_T1_a.csv']</td>
<td style="text-align: left;">['a']</td>
<td style="text-align: right;">10498</td>
<td style="text-align: left;">IgA+</td>
<td style="text-align: left;"></td>
<td style="text-align: right;">nan</td>
<td style="text-align: left;"></td>
<td style="text-align: left;"></td>
</tr>
<tr>
<td style="text-align: right;">4485</td>
<td style="text-align: left;">Sort</td>
<td style="text-align: left;">2022-10-25</td>
<td style="text-align: left;">S6C</td>
<td style="text-align: left;">G003831</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">4</td>
<td style="text-align: left;">V160</td>
<td style="text-align: left;">eODGT8</td>
<td style="text-align: left;">PBMC</td>
<td style="text-align: left;">DV</td>
<td style="text-align: left;">Summary</td>
<td style="text-align: left;">T1</td>
<td style="text-align: left;">P27</td>
<td style="text-align: left;">IgG-IgM-IgD-</td>
<td style="text-align: left;">count</td>
<td style="text-align: left;">.csv</td>
<td style="text-align: left;">['g003/G003/sorting/G003/Sorts/Sort_RunDate221025_UploadDate221101/ClinicalSamples/PopulationSummaryFilesFromDV/Sort_221025_S6C_G003831_V160_eODGT8_PBMC_HT02_DV_Summary_T1_P03_a.csv']</td>
<td style="text-align: left;">['a']</td>
<td style="text-align: right;">100696</td>
<td style="text-align: left;"></td>
<td style="text-align: left;"></td>
<td style="text-align: right;">nan</td>
<td style="text-align: left;">P03</td>
<td style="text-align: left;">HT02</td>
</tr>
<tr>
<td style="text-align: right;">2343</td>
<td style="text-align: left;">PreS</td>
<td style="text-align: left;">2022-11-04</td>
<td style="text-align: left;">S6C</td>
<td style="text-align: left;">G003136</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">-5</td>
<td style="text-align: left;">V091</td>
<td style="text-align: left;">eODGT8</td>
<td style="text-align: left;">PBMC</td>
<td style="text-align: left;">DV</td>
<td style="text-align: left;">Summary</td>
<td style="text-align: left;">T1</td>
<td style="text-align: left;">P12</td>
<td style="text-align: left;">IgD+/Antigen++/KO-</td>
<td style="text-align: left;">count</td>
<td style="text-align: left;">.csv</td>
<td style="text-align: left;">['g003/G003/sorting/G003/Prescreens/Prescreen_RunDate221104_UploadDate221129/PopulationSummaryFilesFromDV/PreS_221104_S6C_G003136_V091_eODGT8_PBMC_DV_Summary_T1_a.csv']</td>
<td style="text-align: left;">['a']</td>
<td style="text-align: right;">47</td>
<td style="text-align: left;">IgD+</td>
<td style="text-align: left;">epitope_pos_igd_pos_b_cells_rev</td>
<td style="text-align: right;">nan</td>
<td style="text-align: left;"></td>
<td style="text-align: left;"></td>
</tr>
</tbody>
</table>
<h2 id="demultiplexing">Demultiplexing<a class="headerlink" href="#demultiplexing" title="Permanent link">&para;</a></h2>
<p>The input to the demultiplexing part of the pipeline will be the sequencing and flow file paths</p>
<div class="tabbed-set tabbed-alternate" data-tabs="2:2"><input checked="checked" id="__tabbed_2_1" name="__tabbed_2" type="radio" /><input id="__tabbed_2_2" name="__tabbed_2" type="radio" /><div class="tabbed-labels"><label for="__tabbed_2_1">:material-console-line: Command Line Usage</label><label for="__tabbed_2_2"> :material-api: Python</label></div>
<div class="tabbed-content">
<div class="tabbed-block">
<p><div class="termy">
<div class="highlight"><pre><span></span><code>$<span class="w"> </span>g00x<span class="w"> </span>g003<span class="w"> </span>pipeline<span class="w"> </span>demultiplex<span class="w"> </span>-o<span class="w"> </span>g003/G003/output/demultiplex<span class="w"> </span>-f<span class="w"> </span>/path/to/flow<span class="w"> </span>-s<span class="w"> </span>/path/to/sequencing
</code></pre></div>
</div></p>
</div>
<div class="tabbed-block">
<div class="highlight"><pre><span></span><code><span class="kn">import</span><span class="w"> </span><span class="nn">pandas</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">pd</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">g00x.data</span><span class="w"> </span><span class="kn">import</span> <span class="n">Data</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">g00x.sequencing.tenX</span><span class="w"> </span><span class="kn">import</span> <span class="n">run_demultiplex</span>

<span class="n">flow_path</span> <span class="o">=</span> <span class="s2">&quot;path/to/flow&quot;</span>
<span class="n">sequencing_path</span> <span class="o">=</span> <span class="s2">&quot;path/to/sequencing&quot;</span>
<span class="n">data</span> <span class="o">=</span> <span class="n">Data</span><span class="p">()</span>
<span class="n">merged_dataframe</span><span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span> <span class="o">=</span> <span class="n">merge_flow_and_sequencing</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">flow_path</span><span class="p">,</span> <span class="n">sequencing_path</span><span class="p">)</span>  <span class="c1"># type: ignore</span>
<span class="n">demultiplex_df</span> <span class="o">=</span> <span class="n">run_demultiplex</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">merged_dataframe</span><span class="p">,</span> <span class="n">out</span><span class="p">,</span> <span class="n">overwrite</span><span class="p">)</span>
<span class="n">demultiplex_df</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="s2">&quot;demultiplex.csv&quot;</span><span class="p">)</span>
<span class="n">demultiplex_df</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="s2">&quot;demultiplex.feather&quot;</span><span class="p">)</span>
</code></pre></div>
</div>
</div>
</div>
<p>The demulitplex algorithm will add the following fields in demultiplex output</p>
<table>
<thead>
<tr>
<th style="text-align: center;">Column</th>
<th style="text-align: left;">Definition</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align: center;">vdj_run_dir</td>
<td style="text-align: left;">The full path to the vdj run directory, e.g. the Illumina directory</td>
</tr>
<tr>
<td style="text-align: center;">cso_run_dir</td>
<td style="text-align: left;">The full path to the cso run directory, e.g. the Illumina directory</td>
</tr>
<tr>
<td style="text-align: center;">vdj_sample_name</td>
<td style="text-align: left;">The unique vdj sample name given to each row</td>
</tr>
<tr>
<td style="text-align: center;">cso_sample_name</td>
<td style="text-align: left;">The unique cso sample name given to each row</td>
</tr>
<tr>
<td style="text-align: center;">vdj_fastq_dir</td>
<td style="text-align: left;">The full path to the vdj fastq directory</td>
</tr>
<tr>
<td style="text-align: center;">cso_fastq_dir</td>
<td style="text-align: left;">The full path to the cso fastq directory</td>
</tr>
</tbody>
</table>
<p>An example demultiplexing output dataframe is found below.</p>
<table>
<thead>
<tr>
<th style="text-align: right;"></th>
<th style="text-align: left;">ptid</th>
<th style="text-align: right;">group</th>
<th style="text-align: right;">weeks</th>
<th style="text-align: left;">visit_id</th>
<th style="text-align: left;">probe_set</th>
<th style="text-align: left;">sample_type</th>
<th style="text-align: left;">run_date</th>
<th style="text-align: left;">sort_pool</th>
<th style="text-align: left;">hashtag</th>
<th style="text-align: left;">run_dir_path</th>
<th style="text-align: left;">pool_number</th>
<th style="text-align: left;">sorted_date</th>
<th style="text-align: right;">vdj_sequencing_replicate</th>
<th style="text-align: right;">cso_sequencing_replicate</th>
<th style="text-align: right;">vdj_lirary_replicate</th>
<th style="text-align: right;">cso_library_replicate</th>
<th style="text-align: right;">bio_replicate</th>
<th style="text-align: left;">vdj_index</th>
<th style="text-align: left;">feature_index</th>
<th style="text-align: left;">vdj_run_id</th>
<th style="text-align: left;">cso_run_id</th>
<th style="text-align: left;">vdj_run_dir_path</th>
<th style="text-align: left;">cso_run_dir_path</th>
<th style="text-align: left;">vdj_fastq_dir</th>
<th style="text-align: left;">vdj_sample_name</th>
<th style="text-align: left;">cso_fastq_dir</th>
<th style="text-align: left;">cso_sample_name</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align: right;">0</td>
<td style="text-align: left;">G003516</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">-5</td>
<td style="text-align: left;">V091</td>
<td style="text-align: left;">eODGT8</td>
<td style="text-align: left;">PBMC</td>
<td style="text-align: left;">2022-09-27</td>
<td style="text-align: left;">P01</td>
<td style="text-align: left;">HT01</td>
<td style="text-align: left;">/mnt/fsx/workspace/jwillis/repos/G00x/g003/G003/sequencing/G003/run0002</td>
<td style="text-align: left;">P01</td>
<td style="text-align: left;">2022-09-27</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: left;">SI-TT-D6</td>
<td style="text-align: left;">SI-TN-D6</td>
<td style="text-align: left;">221006_VH00497_31_AAAVKCLHV</td>
<td style="text-align: left;">221006_VH00497_31_AAAVKCLHV</td>
<td style="text-align: left;">/mnt/fsx/workspace/jwillis/repos/G00x/g003/G003/sequencing/G003/run0002/221006_VH00497_31_AAAVKCLHV</td>
<td style="text-align: left;">/mnt/fsx/workspace/jwillis/repos/G00x/g003/G003/sequencing/G003/run0002/221006_VH00497_31_AAAVKCLHV</td>
<td style="text-align: left;">/mnt/fsx/workspace/jwillis/repos/G00x/g003/G003/sequencing/G003/run0002/working_directory/demultiplexed/29cc0e71cb9200226957921707138c5c/outs/fastq_path</td>
<td style="text-align: left;">vdj-SI-TT-D6</td>
<td style="text-align: left;">/mnt/fsx/workspace/jwillis/repos/G00x/g003/G003/sequencing/G003/run0002/working_directory/demultiplexed/29cc0e71cb9200226957921707138c5c/outs/fastq_path</td>
<td style="text-align: left;">cso-SI-TN-D6</td>
</tr>
<tr>
<td style="text-align: right;">1</td>
<td style="text-align: left;">G003516</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">4</td>
<td style="text-align: left;">V160</td>
<td style="text-align: left;">eODGT8</td>
<td style="text-align: left;">PBMC</td>
<td style="text-align: left;">2022-09-27</td>
<td style="text-align: left;">P01</td>
<td style="text-align: left;">HT02</td>
<td style="text-align: left;">/mnt/fsx/workspace/jwillis/repos/G00x/g003/G003/sequencing/G003/run0002</td>
<td style="text-align: left;">P01</td>
<td style="text-align: left;">2022-09-27</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: left;">SI-TT-D6</td>
<td style="text-align: left;">SI-TN-D6</td>
<td style="text-align: left;">221006_VH00497_31_AAAVKCLHV</td>
<td style="text-align: left;">221006_VH00497_31_AAAVKCLHV</td>
<td style="text-align: left;">/mnt/fsx/workspace/jwillis/repos/G00x/g003/G003/sequencing/G003/run0002/221006_VH00497_31_AAAVKCLHV</td>
<td style="text-align: left;">/mnt/fsx/workspace/jwillis/repos/G00x/g003/G003/sequencing/G003/run0002/221006_VH00497_31_AAAVKCLHV</td>
<td style="text-align: left;">/mnt/fsx/workspace/jwillis/repos/G00x/g003/G003/sequencing/G003/run0002/working_directory/demultiplexed/29cc0e71cb9200226957921707138c5c/outs/fastq_path</td>
<td style="text-align: left;">vdj-SI-TT-D6</td>
<td style="text-align: left;">/mnt/fsx/workspace/jwillis/repos/G00x/g003/G003/sequencing/G003/run0002/working_directory/demultiplexed/29cc0e71cb9200226957921707138c5c/outs/fastq_path</td>
<td style="text-align: left;">cso-SI-TN-D6</td>
</tr>
</tbody>
</table>
<h2 id="vdj">VDJ<a class="headerlink" href="#vdj" title="Permanent link">&para;</a></h2>
<p>The output of demultiplexing pipeline will be used as input, see the <a href="#10x-pipeline">pipeline</a></p>
<div class="tabbed-set tabbed-alternate" data-tabs="3:2"><input checked="checked" id="__tabbed_3_1" name="__tabbed_3" type="radio" /><input id="__tabbed_3_2" name="__tabbed_3" type="radio" /><div class="tabbed-labels"><label for="__tabbed_3_1">:material-console-line: Command Line Usage</label><label for="__tabbed_3_2"> :material-api: Python</label></div>
<div class="tabbed-content">
<div class="tabbed-block">
<p>The following will run the VDJ pipeline from the demultiplex dataframe and output vdj.feather inside of the output folder.</p>
<p><div class="termy">
<div class="highlight"><pre><span></span><code>Run<span class="w"> </span>it<span class="w"> </span>from<span class="w"> </span>demultiplexed<span class="w"> </span>dataframe
$<span class="w"> </span>g00x<span class="w"> </span>g003<span class="w"> </span>pipeline<span class="w"> </span>vdj<span class="w"> </span>-o<span class="w"> </span>g003/G003/output/vdj<span class="w"> </span>-d<span class="w"> </span>output/demultiplexed.feather
</code></pre></div>
</div></p>
</div>
<div class="tabbed-block">
<p>You can run the same with the following python code.</p>
<div class="highlight"><pre><span></span><code><span class="kn">from</span><span class="w"> </span><span class="nn">g00x.sequencing.tenX</span><span class="w"> </span><span class="kn">import</span> <span class="n">run_vdj</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">g00x.data</span><span class="w"> </span><span class="kn">import</span> <span class="n">Data</span>

<span class="n">data</span> <span class="o">=</span> <span class="n">Data</span><span class="p">()</span>
<span class="n">demultiplex_dataframe</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_feather</span><span class="p">(</span><span class="n">demultiplex_dataframe_path</span><span class="p">)</span>
<span class="n">run_vdj</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">demultiplex_dataframe</span><span class="p">,</span> <span class="n">out</span><span class="p">)</span>
</code></pre></div>
</div>
</div>
</div>
<p>The output vdj dataframe will only contain one additional field</p>
<table>
<thead>
<tr>
<th style="text-align: center;">Column</th>
<th style="text-align: left;">Definition</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align: center;">vdj_output</td>
<td style="text-align: left;">The full path to the vdj output folder</td>
</tr>
</tbody>
</table>
<h2 id="cso">CSO<a class="headerlink" href="#cso" title="Permanent link">&para;</a></h2>
<p>This CSO pipeline will run the cellranger count part and output a feature matrix. It also uses the demultiplex.feather as input.</p>
<div class="tabbed-set tabbed-alternate" data-tabs="4:2"><input checked="checked" id="__tabbed_4_1" name="__tabbed_4" type="radio" /><input id="__tabbed_4_2" name="__tabbed_4" type="radio" /><div class="tabbed-labels"><label for="__tabbed_4_1">:material-console-line: Command Line Usage</label><label for="__tabbed_4_2"> :material-api: Python</label></div>
<div class="tabbed-content">
<div class="tabbed-block">
<p>The following will run the CSO pipeline from the demultiplex dataframe and output cso.feather inside of the output folder.</p>
<p><div class="termy">
<div class="highlight"><pre><span></span><code>Run<span class="w"> </span>it<span class="w"> </span>from<span class="w"> </span>demultiplexed<span class="w"> </span>dataframe
$<span class="w"> </span>g00x<span class="w"> </span>g003<span class="w"> </span>pipeline<span class="w"> </span>cso<span class="w"> </span>-o<span class="w"> </span>g003/G003/output/cso<span class="w"> </span>-d<span class="w"> </span>output/demultiplexed.feather
</code></pre></div>
</div></p>
</div>
<div class="tabbed-block">
<p>You can run the same with the following python code.</p>
<div class="highlight"><pre><span></span><code><span class="kn">from</span><span class="w"> </span><span class="nn">g00x.sequencing.tenX</span><span class="w"> </span><span class="kn">import</span> <span class="n">run_cso</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">g00x.data</span><span class="w"> </span><span class="kn">import</span> <span class="n">Data</span>

<span class="n">data</span> <span class="o">=</span> <span class="n">Data</span><span class="p">()</span>
<span class="n">demultiplex_dataframe</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_feather</span><span class="p">(</span><span class="n">demultiplex_dataframe_path</span><span class="p">)</span>
<span class="n">run_cso</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">demultiplex_dataframe</span><span class="p">,</span> <span class="n">out</span><span class="p">)</span>
</code></pre></div>
</div>
</div>
</div>
<p>The output cso dataframe will only contain one additional field</p>
<table>
<thead>
<tr>
<th style="text-align: center;">Column</th>
<th style="text-align: left;">Definition</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align: center;">cso_output</td>
<td style="text-align: left;">The full path to the cso output folder</td>
</tr>
</tbody>
</table>
<h2 id="airr">AIRR<a class="headerlink" href="#airr" title="Permanent link">&para;</a></h2>
<p>The output of the VDJ and CSO can now be combined to get a final sequencing dataframe. This is the final sequencing dataframe that will be used for the analysis.</p>
<p>The AIRR protocol does the following.</p>
<p>It will...</p>
<ol>
<li>Run SADIE AIRR on the VDJ contigs to get a formalized AIRR dataframe.</li>
<li>Analyze the feature barcodes and assign cellids to the correct participant.</li>
<li>Add the PubIDs to the AIRR dataframe.</li>
<li>Run mutational analysis</li>
<li>Run iGL assignment to all sequences</li>
<li>Determine if sequence is VRC01 class</li>
<li>Add mutational sets (find what mutations are VRC01-like)</li>
<li>Cluster sequences</li>
<li>Determine isotype</li>
</ol>
<div class="tabbed-set tabbed-alternate" data-tabs="5:2"><input checked="checked" id="__tabbed_5_1" name="__tabbed_5" type="radio" /><input id="__tabbed_5_2" name="__tabbed_5" type="radio" /><div class="tabbed-labels"><label for="__tabbed_5_1">:material-console-line: Command Line Usage</label><label for="__tabbed_5_2"> :material-api: Python</label></div>
<div class="tabbed-content">
<div class="tabbed-block">
<p>The following will run the AIRR pipeline from the VDJ and CSO dataframes and output airr.feather inside of the output folder.</p>
<p><div class="termy">
<div class="highlight"><pre><span></span><code>Run<span class="w"> </span>it<span class="w"> </span>from<span class="w"> </span>VDJ<span class="w"> </span>and<span class="w"> </span>CSO<span class="w"> </span>dataframes
$<span class="w"> </span>g00x<span class="w"> </span>g003<span class="w"> </span>pipeline<span class="w"> </span>airr<span class="w"> </span>-o<span class="w"> </span>g003/G003/output/airr<span class="w"> </span>-v<span class="w"> </span>output/vdj.feather<span class="w"> </span>-c<span class="w"> </span>output/cso.feather
</code></pre></div>
</div></p>
</div>
<div class="tabbed-block">
<p>You can run the same with the following python code.</p>
<div class="highlight"><pre><span></span><code><span class="kn">from</span><span class="w"> </span><span class="nn">g00x.data</span><span class="w"> </span><span class="kn">import</span> <span class="n">Data</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">g00x.sequencing.airr</span><span class="w"> </span><span class="kn">import</span> <span class="n">run_airr</span>

<span class="n">vdj_out</span> <span class="o">=</span> <span class="s1">&#39;output/vdj.feather&#39;</span>
<span class="n">cso_out</span> <span class="o">=</span> <span class="s1">&#39;output/cso.feather&#39;</span>
<span class="n">data</span> <span class="o">=</span> <span class="n">Data</span><span class="p">()</span>
<span class="n">vdj_dataframe</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_feather</span><span class="p">(</span><span class="n">vdj_out</span><span class="p">)</span>
<span class="n">cso_dataframe</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_feather</span><span class="p">(</span><span class="n">cso_out</span><span class="p">)</span>
<span class="n">output_df</span> <span class="o">=</span> <span class="n">run_airr</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">vdj_dataframe</span><span class="p">,</span> <span class="n">cso_dataframe</span><span class="p">,</span> <span class="n">out</span><span class="p">,</span> <span class="n">overwrite</span><span class="p">)</span>
</code></pre></div>
</div>
</div>
</div>
<p>An output dataframe will take the following:</p>
<p>|       | cellid                                     | pubID    | ptid      | group | weeks | visit_id | probe_set | sample_type | run_date   | sort_pool | hashtag | run_dir_path                                                            | pool_number | sorted_date | vdj_sequencing_replicate | cso_sequencing_replicate | vdj_lirary_replicate | cso_library_replicate | bio_replicate | vdj_index | feature_index | vdj_run_id                  | cso_run_id                  | vdj_run_dir_path                                                                                    | cso_run_dir_path                                                                                    | vdj_fastq_dir                                                                                                                                            | vdj_sample_name | cso_fastq_dir                                                                                                                                            | cso_sample_name | vdj_output                                                                                                    | cso_output                                                                                                    | sadie_airr_path                                                                                                                       | paired_sadie_airr_path                                                                                                                       | cellhash           | sequence_id_heavy           | sequence_heavy                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                       | reference_name_heavy | locus_heavy | stop_codon_heavy | vj_in_frame_heavy | v_frameshift_heavy | productive_heavy | rev_comp_heavy | complete_vdj_heavy | v_call_top_heavy | v_call_heavy | d_call_top_heavy | d_call_heavy | j_call_top_heavy | j_call_heavy | c_call_heavy | sequence_alignment_heavy                                                                                                                                                                                                                                                                                                                                                  | germline_alignment_heavy                                                                                                                                                                                                                                                                                                                                                  | sequence_alignment_aa_heavy                                                                                              | germline_alignment_aa_heavy                                                                                              | v_alignment_start_heavy | v_alignment_end_heavy | d_alignment_start_heavy | d_alignment_end_heavy | j_alignment_start_heavy | j_alignment_end_heavy | c_alignment_start_heavy | c_alignment_end_heavy | v_sequence_alignment_heavy                                                                                                                                                                                                                                                                               | v_sequence_alignment_aa_heavy                                                                      | v_germline_alignment_heavy                                                                                                                                                                                                                                                                               | v_germline_alignment_aa_heavy                                                                      | d_sequence_alignment_heavy | d_sequence_alignment_aa_heavy | d_germline_alignment_heavy | d_germline_alignment_|</p>













              </article>
            </div>


  <script>var tabs=__md_get("__tabs");if(Array.isArray(tabs))e:for(var set of document.querySelectorAll(".tabbed-set")){var labels=set.querySelector(".tabbed-labels");for(var tab of tabs)for(var label of labels.getElementsByTagName("label"))if(label.innerText.trim()===tab){var input=document.getElementById(label.htmlFor);input.checked=!0;continue e}}</script>

<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>

      </main>

        <footer class="md-footer">

  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">


    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>

</div>

        <div class="md-social">








    <a href="https://github.com/schieflab/g00x" target="_blank" rel="noopener" title="github.com" class="md-social__link">
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 480 512"><!--! Font Awesome Free 6.7.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M186.1 328.7c0 20.9-10.9 55.1-36.7 55.1s-36.7-34.2-36.7-55.1 10.9-55.1 36.7-55.1 36.7 34.2 36.7 55.1M480 278.2c0 31.9-3.2 65.7-17.5 95-37.9 76.6-142.1 74.8-216.7 74.8-75.8 0-186.2 2.7-225.6-74.8-14.6-29-20.2-63.1-20.2-95 0-41.9 13.9-81.5 41.5-113.6-5.2-15.8-7.7-32.4-7.7-48.8 0-21.5 4.9-32.3 14.6-51.8 45.3 0 74.3 9 108.8 36 29-6.9 58.8-10 88.7-10 27 0 54.2 2.9 80.4 9.2 34-26.7 63-35.2 107.8-35.2 9.8 19.5 14.6 30.3 14.6 51.8 0 16.4-2.6 32.7-7.7 48.2 27.5 32.4 39 72.3 39 114.2m-64.3 50.5c0-43.9-26.7-82.6-73.5-82.6-18.9 0-37 3.4-56 6-14.9 2.3-29.8 3.2-45.1 3.2-15.2 0-30.1-.9-45.1-3.2-18.7-2.6-37-6-56-6-46.8 0-73.5 38.7-73.5 82.6 0 87.8 80.4 101.3 150.4 101.3h48.2c70.3 0 150.6-13.4 150.6-101.3m-82.6-55.1c-25.8 0-36.7 34.2-36.7 55.1s10.9 55.1 36.7 55.1 36.7-34.2 36.7-55.1-10.9-55.1-36.7-55.1"/></svg>
    </a>

</div>

    </div>
  </div>
</footer>

    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>




      <script id="__config" type="application/json">{"base": "..", "features": ["search.suggest", "search.highlight", "content.tabs.link", "search.share", "navigation.tracking", "navigation.tabs.sticky"], "search": "../assets/javascripts/workers/search.f8cc74c7.min.js", "tags": null, "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}, "version": null}</script>


      <script src="../assets/javascripts/bundle.c8b220af.min.js"></script>

        <script src="https://unpkg.com/mermaid@8.4.6/dist/mermaid.min.js"></script>

        <script src="../js/termynal.js"></script>

        <script src="../js/custom.js"></script>


  </body>
</html>
